File: HW6/LinkedList.cpp
#include "LinkedList.h"
#include <iostream>
#include <algorithm>

namespace GTU {

    //Constructors

    template<typename T, typename C>
    LinkedList<T, C>::LinkedList(void) {
        /* INTENTIONALLY EMPTY */
    }

    template<typename T, typename C>
    LinkedList<T, C>::LinkedList(C &c) {
        auto it = c.begin();
        for (int i = 0; i < c.size(); ++i) {
            data.insert(data.end(), *(it));
            ++it;
        }
        this->sortList();

    }

    template<typename T, typename C>
    LinkedList<T, C>::LinkedList(int initialCapacity) {
        /* INTENTIONALLY EMPTY */
    }

    //Returns the value at the begin of this collection

    template<typename T, typename C>
    T LinkedList<T, C>::element(){
        auto i = data.begin();
        return *i;
    }

    //Adds given value to end of this collection

    template<typename T, typename C>
    bool LinkedList<T, C>::offer(T e){
        this->add(e);
        return 0;
    }

    //Removes the head

    template<typename T, typename C>
    T LinkedList<T, C>::poll(void)throw(std::out_of_range){
        if(this->isEmpty()){
            throw std::out_of_range("This collection is empty!!\n");
        } else {
            auto i = data.begin();
            data.erase(i);
        }
        return *(data.begin());
    }

    //This function creates an Iterator for collection

    template<typename T, typename C>
    Iterator<T, C> LinkedList<T, C>::iterator(void) {
        //Iterator<T, C> it = Iterator<T,C>(data);
        return Iterator<T, C>(data);
    }

    //Adds given element to the end of the collection

    template <typename T, typename C>
    bool LinkedList<T, C>::add(T e){

        data.insert(data.end(), e);
        this->sortList();

        //if(emptyCapacity == 0) {
        //} else{
        //   data[data.size() - emptyCapacity] = e;
        //   --emptyCapacity;
        // }
        return true;
    }

    //Adds all of given collection to the end of this collection

    template <typename T, typename C>
    bool LinkedList<T, C >::addAll(Collection<T, C> &c){
        for(int i = 0 ; i < c.size() ; ++i) {
            this->add(c[i]);
        }
        return true;
        /*auto i = c.iterator();

        if(!(c.isEmpty())) {

            while (i.hasNext()) {
                this->add(i.elementgetter());
                i.next();
            }
            this->add(i.elementgetter());
        }
        return true;*/
    }

    //Makes size of this collection empty

    template <typename T, typename C>
    void LinkedList<T, C>::clear(){
        data.clear();
    }

    //Checks if given element exist in this collection

    template <typename T, typename C>
    bool LinkedList<T, C>::contains(T e){
        for(auto i = data.begin(); i != data.end(); ++i){
            if( e == *i){
                return true;
            }
        }

        return false;
    }

    //Checks if all of given collection exists in this collection

    template <typename T, typename C>
    bool LinkedList<T, C>::containsAll(Collection<T, C > &c) {

        for(int i = 0; i < c.size(); ++i){
            if(!(this->contains(c[i]))) {
                return false;
            }
        }

        return true;
/*  //Alternative code
               auto i = c.iterator();

        if(!(c.isEmpty())) {

            while (i.hasNext()) {
                if(this->contains(i.elementgetter())) {

                    return false;
                }
                i.next();
            }
            if(this->contains(i.elementgetter())) {
                return false;
            }
        }*/
    }

    //Checks if this collection is empty

    template<typename T, typename C>
    bool LinkedList<T, C>::isEmpty(void) {
        if (data.size() == 0) {
            return true;
        } else {
            return false;
        }
    }

    //Removes the given element if it exists in this collection

    template <typename T, typename C>
    bool LinkedList<T, C>::remove(T e)throw(std::invalid_argument) {
        if(this->contains(e)){
            auto i = data.begin();
            if(e == *i) {
                data.erase(i);
                return true;
            } else{
                throw std::invalid_argument("This element can not be removed because this collection is FIFO\n");
            }
        } else {
            throw std::invalid_argument("This element does not exist!!\n");
        }
    }

    //Removes all of the given element if it exists in this collection

    template <typename T, typename C>
    bool LinkedList<T, C>::removeAll(Collection<T, C> &c){
  /*      for(int i = 0; i < c.size(); ++i){
            this->remove(c[i]);
        }
*/
  //I use iterator to make sure that Iterator it works
        auto i = c.iterator();

        if(!(c.isEmpty())) {

            while (i.hasNext()) {
                remove(i.elementgetter());
                i.next();
            }
            this->remove(i.elementgetter());
        }
        return true;
    }

    //Returns intersection

    template <typename T, typename C>
    LinkedList<T, C>& LinkedList<T, C>::retainAll(Collection<T, C > &c){
        int j;
        auto it = c.iterator();
        static LinkedList<T, C> temp;
		//To make temp empty (it is static) i call clear function
		temp.clear();
        if(!(c.isEmpty())) {
            for (auto i = data.begin(); i != data.end(); ++i) {
                if(c.contains(*i)){
                    temp.add(*i);
                }
            }
        }
        return temp;
    }

    //Returns size of this collection

    template <typename T, typename C>
    int LinkedList<T, C>::size(void){
        return data.size();
    }

    //[] overloading to reach elements of this collection

    template <typename T, typename C>
    T LinkedList<T, C>::operator [](int i){
        auto it = data.begin();
        for( int j = 0; j < i; ++j ){
            ++it;
        }
        return *it;
    }

    //Sorts lists

    template <typename T, typename C>
    void LinkedList<T, C>::sortList(void){
        std::vector<T> temp;

        for (auto i = data.begin(); i != data.end(); ++i){
            temp.push_back(*i);
        }

        data.clear();

        std::sort(temp.begin(), temp.end());

        for (auto i = 0 ; i < temp.size() ; ++i){
            data.insert(data.end(), temp[i]);
        }    }
}

File: HW8/171044053_CSE241_HW8/src/LinkedList.java
/**
 * Implemets List and Queue methods 
 * LinkedList is generic
 * @author İlkay CAN - 171044053
 *
 * @param <E> defines type of data
 */

import java.util.Arrays;


public class LinkedList<E> implements List<E>, Queue<E> {
	
	private E[] data;

	//Constructor
	
	/**
	 * Constructor
	 */
	@SuppressWarnings("unchecked")
	public LinkedList(){
		data = (E[])new Object[0];
	}
	
    //This function creates an Iterator for collection
	public Iterator <E> iterator(){
		return new Iterator<E>(data);
	}
	
    //Adds given element to the end of the collection
	@SuppressWarnings("unchecked")
	public boolean add (E e){
		E temp[] = (E[])new Object[data.length + 1];
        System.arraycopy(data, 0, temp, 0, data.length);
        temp[data.length] = e;
        data = temp;

        Arrays.sort(data);
        	
		return true;
	}
	
    //Adds all of given collection to the end of this collection
	public boolean addAll(Collection<E> c){

		for(int i = 0 ; i < c.size() ; ++i) {
            this.add(c.at(i));
        }
        return true;	
    }
	
    //Makes collection empty
	@SuppressWarnings("unchecked")
	public void clear(){
    	E temp[] = (E[])new Object[0];
    	data = temp;
	}
	
    //Checks if given element exist in this collection
	public boolean contains(E e){
		for(E i : data){
            if( e == i){
                return true;
            }
        }

        return false;
	}
	
    //Checks if all of given collection exists in this collection
	public boolean containsAll(Collection<E> c){
        for(int i = 0; i < c.size(); ++i){
            if(!(this.contains(c.at(i)))) {
                return false;
            }
        }

        return true;
	}
	
    //Checks if this collection is empty
	public boolean isEmpty(){
        if (data.length == 0) {
            return true;
        } else {
            return false;
        }
	}
    //Removes the given element if it exists in this collection
	
	/**
	 * removes given element 
	 * @param e
	 * @return true if remove is succesfull
	 * @throws IllegalArgumentException if given element does not exist or given element is not at head
	 */
	@SuppressWarnings("unchecked")
	public boolean remove(E e) throws IllegalArgumentException{
	    if(this.contains(e)){
	        if(e == data[0]) {
		    	E temp[] = (E[])new Object[data.length - 1];
	        	System.arraycopy(data, 1, temp, 0, temp.length);
	        	data = temp;
	        } else{
	            throw new IllegalArgumentException("This element can not be removed because this collection is FIFO\n");
	        }
	    } else {
	        throw new IllegalArgumentException("This element does not exist!!\n");
	    }
	    return true;
	}
	
    //Removes all of the given element if it exists in this collection
	public boolean removeAll(Collection<E> c){
        for(int i = 0; i < c.size(); ++i){
            this.remove(c.at(i));
        }
        return true;
	}
	
    //Returns intersection
	public Collection<E> retainAll( Collection<E> c){

	      LinkedList<E> temp = new LinkedList<E>();

	      if(!(c.isEmpty())){
	    	  for(E i : data){
	    		  if(c.contains(i)){
	    			  temp.add(i);
	    		  }
	    	  }
	      }
	     
	      return temp;
	}
	
    //Returns size of this collection
	public int size(){
		return data.length;
	}
	
	//returns head
	
	/**
	 * 
	 * @return element in first index
	 * @throws IndexOutOfBoundsException if Collection is empty
	 */
	
	public E element() throws IndexOutOfBoundsException{
		if(data.length != 0){
			return data[0];
		} else {
			throw new IndexOutOfBoundsException("This colleciton is empty");
		}
	}
	
	//Adds given value to Collection
	public boolean offer(E e){
        this.add(e);
  		return true;
	}
	
	//Removes last element
	
	/**
	 * removes last element
	 * @return last element
	 * @throws IndexOutOfBoundsException if Collection is already empty
	 */
	public E poll()throws IndexOutOfBoundsException{
        if(this.isEmpty()){
            throw new IndexOutOfBoundsException("This collection is empty!!\n");
        } else {
        	this.remove(data[0]);
        }
        return data[0];
	}
	
	//returns element in the given index
	public E at(int index){
		return data[index];
	}
}
